### ISSUE TITLE
Pavucontrol is shown in input apps

### ISSUE THREAD

**Issue by Digitalone1** [wwmm/easyeffects#1128]
I rarely touch the stream input pipeline, but today I found out this:

![Schermata del 2021-09-03 18-58-58](https://user-images.githubusercontent.com/25790525/132041963-adee350f-48a6-400a-abc2-d8e51717748c.png)

Wasn't it supposed to be hard blocklisted? 

@wwmm Am I wrong or this was happening also in PulseEffects because the localized app name skipped the blocklist and you had to exclude apps on other parameters?

**Comment by wwmm** [comment]
>Wasn't it supposed to be hard blocklisted?

I tried it on my computer and there was no entry. But I am not using translations. My window is in English. I think that the string currently used in our built-in blocklist may be translation dependent. IF that is the case we will have to replace it by another that does not change in this situation.

**Comment by Digitalone1** [comment]
That is the case, we have to find a parameter reporting the application name, not the translated title.

**Comment by wwmm** [comment]
The command `pw-cli list-objects Node` will help to find other tags that may not have been translated.

**Comment by wwmm** [comment]
The output for pavucontrol is
```
                id 125, type PipeWire:Interface:Node/3
 		factory.id = "6"
 		client.id = "206"
 		application.name = "PulseAudio Volume Control"
 		node.name = "PulseAudio Volume Control"
 		media.class = "Stream/Input/Audio"
```

At this moment we are checking the `node.name` 
https://github.com/wwmm/easyeffects/blob/e7597e192c1cee2db1f30500c2fa3c5a5ef875fe/src/pipe_manager.cpp#L797
https://github.com/wwmm/easyeffects/blob/e7597e192c1cee2db1f30500c2fa3c5a5ef875fe/include/pipe_manager.hpp#L193

I wonder if `application.name` is not translated.

**Comment by Digitalone1** [comment]
```
	id 54, type PipeWire:Interface:Node/3
 		factory.id = "6"
 		client.id = "51"
 		application.name = "Regolazione del volume PulseAudio"
 		node.name = "Regolazione del volume PulseAudio"
 		media.class = "Stream/Input/Audio"
	id 56, type PipeWire:Interface:Node/3
 		factory.id = "6"
 		client.id = "51"
 		application.name = "Regolazione del volume PulseAudio"
 		node.name = "Regolazione del volume PulseAudio"
 		media.class = "Stream/Input/Audio"
```

So we can't blocklist it. :(

**Comment by wwmm** [comment]
That is concerning... I think most of the applications do not have these tags translated. I wonder why they decided to do that in Pavucontrol...

**Comment by Digitalone1** [comment]
I'm trying with regex.

**Comment by wwmm** [comment]
I think that there is another property we can use. Once you find the id of the Pavucontrol node run `pw-cli info id_number`. The output will be something like
```
	id: 129
	permissions: rwxm
	type: PipeWire:Interface:Node/3
*	input ports: 2/64
*	output ports: 2/0
*	state: "suspended"
*	properties:
*		client.api = "pipewire-pulse"
*		pulse.server.type = "unix"
*		application.name = "PulseAudio Volume Control"
*		application.id = "org.PulseAudio.pavucontrol"
*		application.icon-name = "audio-card"
*		application.version = "5.0"
*		application.process.id = "2612"
*		application.process.user = "wallace"
*		application.process.host = "wwmm"
*		application.process.binary = "pavucontrol"
*		application.language = "en_US.UTF-8"
*		window.x11.display = ":0"
*		application.process.machine-id = "b3d293e2d17b4948b715903eea76b85f"
*		media.name = "Peak detect"
*		node.rate = "1/25"
*		stream.monitor = "true"
*		node.target = "40"
*		stream.is-live = "true"
*		node.name = "PulseAudio Volume Control"
*		node.autoconnect = "true"
*		node.dont-reconnect = "true"
*		media.class = "Stream/Input/Audio"
*		resample.peaks = "true"
*		channelmix.normalize = "true"
*		adapt.follower.node = ""
*		object.register = "false"
*		factory.id = "6"
*		audio.adapt.follower = ""
*		factory.mode = "merge"
*		library.name = "audioconvert/libspa-audioconvert"
*		client.id = "130"
*		object.id = "129"
*		node.latency = "1/25"
*		pulse.attr.maxlength = "4194304"
*		pulse.attr.fragsize = "4"
*	params: (8)
*	  3 (Spa:Enum:ParamId:EnumFormat) r-
*	  1 (Spa:Enum:ParamId:PropInfo) r-
*	  2 (Spa:Enum:ParamId:Props) rw
*	  4 (Spa:Enum:ParamId:Format) rw
*	  10 (Spa:Enum:ParamId:EnumPortConfig) r-
*	  11 (Spa:Enum:ParamId:PortConfig) rw
*	  15 (Spa:Enum:ParamId:Latency) -w
*	  16 (Spa:Enum:ParamId:ProcessLatency) rw
```

The interesting line is `application.id = "org.PulseAudio.pavucontrol"`. Just like us Pavucontrol defines a unique id. This is not translated. So I think that what we can do is creating another list named like `blocklist_application_id` to be checked in the same place where `blocklist_node_name` is.

**Comment by Digitalone1** [comment]
OK, so no need of regex. This is something you have to do, I have no idea how to get the application id.

I wonder if all blocklisted apps should be based on app id and not only pavucontrol.

**Comment by wwmm** [comment]
>OK, so no need of regex. This is something you have to do, I have no idea how to get the application id.

Ok. Once your new pull request is merged I will do that.

>I wonder if all blocklisted apps should be based on app id and not only pavucontrol.

In an ideal world every Pulseaudio/PipeWire/GTK app developer would define an unique id. Unfortunately we are not in an ideal world... Most developers do not define this tag...

**Comment by wwmm** [comment]
Sometimes PipeWire does not make my life easy... `PW_KEY_APP_ID` is defined when updating node information and not when we receive it for the first time... The problem is by the time we receive this update we have already added the node to our list and emitted the creation signal... So we would have to remove its node after... Not nice...

I will see if `application.process.binary` is a better option.

**Comment by Digitalone1** [comment]
If it's not possible, there's always the regex option. Check if `pulseaudio` and `volume` are both contained in the name.

**Comment by wwmm** [comment]
>I will see if application.process.binary is a better option.

It is also undefined when we receive the node for the first time.  I think that of all the possible `PW_KEY_APP_` only `PW_KEY_APP_NAME` is defined from the start... 

>If it's not possible, there's always the regex option. Check if pulseaudio and volume are both contained in the name.

But will this be a solution to every language? `pulseaudio` will probably be there but I am not so sure people won't  replace `volume` by something else.

**Comment by wwmm** [comment]
The user still can blocklist pavucontrol on his own. And as that list will use the translation it will work. So I am not sure it is worth the added overhead of checking regex or the added complexity of removing the node after adding it. We would also have to kill the proxies that are created to listen to node events...

**Comment by Digitalone1** [comment]
> > I will see if application.process.binary is a better option.
> 
> It is also undefined when we receive the node for the first time. I think that of all the possible `PW_KEY_APP_` only `PW_KEY_APP_NAME` is defined from the start...
> 
> > If it's not possible, there's always the regex option. Check if pulseaudio and volume are both contained in the name.
> 
> But will this be a solution to all the languages? `pulseaudio` will probably be there but I am not so sure people won't replace `volume` by something else.

With most of the languages will work. Anyway it's better if you open an issue on Pipewire Gitlab and ask to set the application id from the very beginning.

The regex solution would be only a partial workaround. If you don't want it, I won't add it and insert the name in the user blocklist.

**Comment by wwmm** [comment]
>Anyway it's better if you open an issue on Pipewire Gitlab and ask to set the application id from the very beginning.

It is something to consider. But in the past I faced other situations where I wanted the information from the start and after looking at PipeWire's code it did not seem that they could do that. I think that it depends on the stream initialization. Some of the information is probably not available from the start even to the server because the stream can only provide it a later time.

>The regex solution would be only a partial workaround. If you don't want it, I won't add it and insert the name in the user blocklist.

For now let's keep things as they are. I will give more thought to this situation.

**Comment by wwmm** [comment]
Maybe we can blocklist it in a indirect way https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/keys.h#L269. `PW_KEY_STREAM_MONITOR` I do not think it makes sense to apply effects to monitor streams. One way could be to blocklist all of them. What would also block pavucontrol streams.

**Comment by Digitalone1** [comment]
> Maybe we can blocklist it in a indirect way https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/keys.h#L269 through `PW_KEY_STREAM_MONITOR`. I do not think it makes sense to apply effects to monitor streams. One way could be to blocklist all of them. What would also block pavucontrol streams.

Sure, this seems a better solution that would also remove pavucontrol from the blocklist array.

**Comment by wwmm** [comment]
`PW_KEY_STREAM_MONITOR` is also undefined when we receive the node for the first time... Unbelievable...

**Comment by Digitalone1** [comment]
I wonder why Pulseaudio could get the app id in its stream and Pipewire can't.

**Comment by wwmm** [comment]
>I wonder why Pulseaudio could get the app id in its stream and Pipewire can't.

Technically speaking it can. But not at the moment I would like it to have this information. As its design is very different from Pulseaudio it is reasonable things are done in different stages.  This is one of the few situations where Pulseaudio's approach is more convenient for us.

**Comment by Digitalone1** [comment]
And if we check the `PW_KEY_STREAM_MONITOR` state in `on_node_info`

https://github.com/wwmm/easyeffects/blob/46fd5182b8e59d221d48eb900a4b68b8c9f2d90f/src/pipe_manager.cpp#L175

and we invoke the `stream_output_removed`/`stream_intput_removed` signal?

Do you think it's possible? The app would be added and soon removed, but it's better than nothing at the actual state.

**Comment by wwmm** [comment]
>Do you think it's possible? The app would be added and soon removed, but it's better than nothing at the actual state.

It is possible but that kind of solution would need more than just blocklisting pavucontrol to justify it. The reason is that calling `stream_*_removed` is not enough. At the lines 823,824 and 826 we add listeners to those nodes that we will have to remove if we are not monitoring them and we also have to remove them from the node list https://github.com/wwmm/easyeffects/blob/46fd5182b8e59d221d48eb900a4b68b8c9f2d90f/src/pipe_manager.cpp#L823. I am not sure I want to be doing these kind of operations inside the `on_node_info` callback. I am also not 100% sure we can just call `spa_hook_remove` to remove a listener inside its own callback. It may be fine. It may not be.

If the user had no way to get rid of Pavucontrol in the applications list it would make sense. But with a custom blocklist available on the user side that can handle this situation is a little questionable to take suck approaches. I will keep thinking about this. Maybe a simpler solution comes to my mind.
